File: binary-search-tree/binary_search_tree.php
<?php
class Node {
    public $data;
    public $left;
    public $right;
    
    public function __construct($data) {
        $this->data = $data;
        $this->left = null;
        $this->right = null;
    }
}

class BST {
    private $root;
    
    public function __construct() {
        $this->root = null;
    }
    
    public function insert($data) {
        $this->root = $this->insertNode($this->root, $data);
    }
    
    private function insertNode($node, $data) {
        if ($node === null) {
            return new Node($data);
        }
        
        if ($data < $node->data) {
            $node->left = $this->insertNode($node->left, $data);
        } else {
            $node->right = $this->insertNode($node->right, $data);
        }
        
        return $node;
    }
    
    public function search($data) {
        return $this->searchNode($this->root, $data);
    }
    
    private function searchNode($node, $data) {
        if ($node === null || $node->data === $data) {
            return $node;
        }
        
        if ($data < $node->data) {
            return $this->searchNode($node->left, $data);
        }
        
        return $this->searchNode($node->right, $data);
    }
    
    public function inorder() {
        $this->inorderTraversal($this->root);
    }
    
    private function inorderTraversal($node) {
        if ($node !== null) {
            $this->inorderTraversal($node->left);
            echo $node->data . " ";
            $this->inorderTraversal($node->right);
        }
    }
}

$bst = new BST();
$values = [50, 30, 70, 20, 40];
foreach ($values as $value) {
    $bst->insert($value);
}
$bst->inorder();
echo "\nSearch 40: " . ($bst->search(40) ? "Found" : "Not found") . "\n";
?>

File: binary-search-tree/binary_search_tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
} Node;

Node* createNode(int data) {
    Node* node = malloc(sizeof(Node));
    node->data = data;
    node->left = node->right = NULL;
    return node;
}

Node* insert(Node* root, int data) {
    if (!root) return createNode(data);
    if (data < root->data) root->left = insert(root->left, data);
    else root->right = insert(root->right, data);
    return root;
}

Node* search(Node* root, int data) {
    if (!root || root->data == data) return root;
    return data < root->data ? search(root->left, data) : search(root->right, data);
}

void inorder(Node* root) {
    if (root) {
        inorder(root->left);
        printf("%d ", root->data);
        inorder(root->right);
    }
}

int main() {
    Node* root = NULL;
    int values[] = {50, 30, 70, 20, 40};
    for (int i = 0; i < 5; i++) root = insert(root, values[i]);
    inorder(root);
    printf("\nSearch 40: %s\n", search(root, 40) ? "Found" : "Not found");
    return 0;
}